pkg/scalers/gcp_pub_sub_scaler.go

```diff
@@ -135,40 +165,73 @@ func (s *pubsubScaler) GetMetricSpecForScaling(context.Context) []v2beta2.Metric
 
 // GetMetrics connects to Stack Driver and finds the size of the pub sub subscription
 func (s *pubsubScaler) GetMetrics(ctx context.Context, metricName string, metricSelector labels.Selector) ([]external_metrics.ExternalMetricValue, error) {
-	size, err := s.GetSubscriptionSize(ctx)
+	var value int64
+	var err error
 
-	if err != nil {
-		gcpPubSubLog.Error(err, "error getting subscription size")
-		return []external_metrics.ExternalMetricValue{}, err
+	if s.metadata.mode == pubsubModeSubscriptionSize {
+		value, err = s.getSubscriptionSize(ctx)
+		if err != nil {
+			gcpPubSubLog.Error(err, "error getting subscription size")
+			return []external_metrics.ExternalMetricValue{}, err
+		}
+	} else if s.metadata.mode == pubsubModeOldestUnackedMessageAge {
+		value, err = s.getOldestUnackedMessageAge(ctx)
+		if err != nil {
+			gcpPubSubLog.Error(err, "error getting oldest unacked message age")
+			return []external_metrics.ExternalMetricValue{}, err
+		}
 	}
 
 	metric := external_metrics.ExternalMetricValue{
 		MetricName: metricName,
-		Value:      *resource.NewQuantity(size, resource.DecimalSI),
+		Value:      *resource.NewQuantity(value, resource.DecimalSI),
 		Timestamp:  metav1.Now(),
 	}
 
 	return append([]external_metrics.ExternalMetricValue{}, metric), nil
 }
 
-// GetSubscriptionSize gets the number of messages in a subscription by calling the
+func (s *pubsubScaler) setStackdriverClient(ctx context.Context) error {
+	var client *StackDriverClient
+	var err error
+	if s.metadata.gcpAuthorization.podIdentityProviderEnabled {
+		client, err = NewStackDriverClientPodIdentity(ctx)
+	} else {
+		client, err = NewStackDriverClient(ctx, s.metadata.gcpAuthorization.GoogleApplicationCredentials)
+	}
+
+	if err != nil {
+		return err
+	}
+	s.client = client
+	return nil
+}
+
+// getSubscriptionSize gets the number of messages in a subscription by calling the
 // Stackdriver api
-func (s *pubsubScaler) GetSubscriptionSize(ctx context.Context) (int64, error) {
+func (s *pubsubScaler) getSubscriptionSize(ctx context.Context) (int64, error) {
 	if s.client == nil {
-		var client *StackDriverClient
-		var err error
-		if s.metadata.gcpAuthorization.podIdentityProviderEnabled {
-			client, err = NewStackDriverClientPodIdentity(ctx)
-		} else {
-			client, err = NewStackDriverClient(ctx, s.metadata.gcpAuthorization.GoogleApplicationCredentials)
+		err := s.setStackdriverClient(ctx)
+		if err != nil {
+			return -1, err
 		}
+	}
+
+	filter := `metric.type="` + pubSubStackDriverSubscriptionSizeMetricName + `" AND resource.labels.subscription_id="` + s.metadata.subscriptionName + `"`
+
+	return s.client.GetMetrics(ctx, filter)
+}
+
+// getOldestUnackedMessageAge gets oldest unacked message age in a subscription by calling stackdriver api
```

Comment:
It could make sense, in order not not repeat this every time you add another metric, to add a generic method to get the metric, and use the metric name as a value in the method or as a class parameter, or something similar.